### PR TITLE
Get unit registry from application

### PR DESCRIPTION
**Pull request by nollety**
# Description

Allow applications using Eradiate to define their own unit registry that Eradiate can fetch (thanks to [pint.get_application_registry](https://pint.readthedocs.io/en/stable/api/base.html#pint.get_application_registry)), modify and use.

Example 1 (the application defines a unit registry):
```python
# my application using Eradiate
import pint

ureg = pint.UnitRegistry()
ureg.define("tatami = 1.8 meter")
pint.set_application_registry()

x = 1 * ureg.tatami 

from eradiate import unit_registry  # unit_registry is the same registry as 'ureg'
y = 2 * unit_registry.meter

# since 'ureg' and 'unit_registry' are the same registries, 'x' and 'y' can be compared
print(y - x)  # 0.2 meter
```
Example 2 (the application defines a unit registry but does not set it as application registry):
```python
# my application using Eradiate
import pint

ureg = pint.UnitRegistry()
ureg.define("tatami = 1.8 meter")

x = 1 * ureg.tatami 

from eradiate import unit_registry  # 'unit_registry' is different from 'ureg'
y = 2 * unit_registry.meter

# since 'ureg' and 'unit_registry' are not the same registries, 'x' and 'y' can not be compared
print(y - x)  # raises
```

Example 3 (the application does not define a unit registry):
```python
# my application using Eradiate
from eradiate import unit_registry  # Eradiate updates pint's default unit registry with its custom units 
```

## Explanation

Eradiate will attempt to modify the application registry to add its custom units (so far, they are `dobson_unit` and `atmo_centimeter`). If the custom unit definition does not already exist, it is added; else, a warning is logged (assuming that the existing unit definition is the same as Eradiate's).
If the application does not define a unit registry, `pint.get_application_registry` is going to return the default `pint` unit registry, to which Eradiate can add its custom units.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I updated the change log if relevant
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license
- [ ] update units guide

### CODE REVIEW

File: src/eradiate/units.py [hunk 1]
```diff
@@ -19,27 +20,33 @@
...
-unit_registry.define(
-    "dobson_unit = 2.687e20 * meter^-2 " "= du = dobson = dobson_units"  # aliases
-)
-"""IUPAC. Compendium of Chemical Terminology, 2nd ed. (the "Gold Book").
-Compiled by A. D. McNaught and A. Wilkinson. Blackwell Scientific Publications,
-Oxford (1997). Online version (2019-) created by S. J. Chalk. ISBN 0-9678550-9-8.
-https://doi.org/10.1351/goldbook."""
-
-unit_registry.define(
+unit_registry = pint.get_application_registry()
```
[the remark this answers is not in the document]
Comment (nollety) [reply src/eradiate/units.py:29]: I am not sure that I understand what you mean with "reset it", but that would be a misuse of the unit registry in that case, I suppose

[the remark this answers is not in the document]
Comment (leroyvn) [reply src/eradiate/units.py:29]: If, for some reason (say, for instance, that someone has a custom registry class they want to use), somebody uses `pint.set_application_registry()`, then `eradiate.unit_registry` won't reflect this change.

**Comment by nollety** [comment]
> are we sure that it's the right choice, i.e. that Eradiate should always share its registry?

Phrased like that, it is difficult to answer yes ;) I don't want to claim that this is 'always' going to be the right solution. At least, from my understanding of pint and Eradiate foreseen usage, it seems to me that it would be convenient for Eradiate' unit registry to be the same as the application registry. This is a suggestion, if you have other ideas of Eradiate usage than illustrated in the above examples, then this could not be appropriate.

File: src/eradiate/units.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nollety) [reply src/eradiate/units.py:29]: Like in above example 1 but with the call to `pint.set_application_registry()` after the statement `from eradiate import unit_registry`?

**Comment by nollety** [comment]
According to pint documentation, it is advised to have only one (that is, shared) unit registry (https://pint.readthedocs.io/en/stable/getting/pint-in-your-projects.html)

File: src/eradiate/units.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (leroyvn) [reply src/eradiate/units.py:29]: Example:
```python
import pint
import eradiate

ureg = MyUnitRegistry()
pint.set_application_registry(ureg)
1.0 * ureg.m == 1.0 * eradiate.unit_registry  # raises
```

Note that the following would most likely fix it:
```python
import pint

ureg = MyUnitRegistry()
pint.set_application_registry(ureg)

import eradiate  # Import behaviour depends on Python global state, it's bad

1.0 * ureg.m == 1.0 * eradiate.unit_registry  # works
```
but IMO being dependent on import order for this to work is not nice.

Lazy registry loading would make snippet 1 work, at the cost of resolving the application registry at every call to `eradiate.unit_registry`. And, like I said, it still doesn't solve the problem of updating static constants instantiated with Pint's default registry.

The bottom line is: if all pint-dependent packages use the application registry, things work as they are expected to. As soon as one (Eradiate, another package or the client code) uses another registry, it becomes fragile at best, and most likely breaks pretty much just like it currently does.

**Comment by leroyvn** [comment]
The doc page recommends doing exactly what I find harmful: to instantiate a registry manually, then declare it as the application registry.

File: src/eradiate/units.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nollety) [reply src/eradiate/units.py:29]: Exactly, this is why I ordered the import statements in the above examples

**Comment by leroyvn** [comment]
I don't see a nicer solution a.t.m. that won't ridiculously complicate the unit registry fetch logic and/or add unnecessary overhead, so let's proceed with that. It's the least bad solution Pint currently offers.

**Comment by nollety** [comment]
> The doc page recommends doing exactly what I find harmful: to instantiate a registry manually, then declare it as the application registry.

The harm being the import statements order?

**Comment by nollety** [comment]
> I don't see a nicer solution a.t.m. that won't ridiculously complicate the unit registry fetch logic and/or add unnecessary overhead, so let's proceed with that. It's the least bad solution Pint currently offers. Let's summarize the limitations here:
> 
> Eradiate uses Pint's current application registry at import time. This means that:
> 
> * A custom application registry must be declared before issuing `import eradiate`.
> * Subsequent modifications of the application registry using `pint.set_application_registry()` will not propagate to Eradiate and will break Eradiate's unit support.

We could add this to the existing units guide.

**Comment by nollety** [comment]
In the case where the application want to declare its custom units (in below example, the `tatami` unit), it is also possible for them to do:

```python
import pint
from eradiate import unit_registry as ureg

# application custom units:
ureg.define("tatami = 1.8 meter")

x = 1 * ureg.tatami 
y = 2 * ureg.meter

print(y - x)  # 0.2 meter
```
or even using a `try`-`catch` to avoid `pint.RedefinitionError`.

**Comment by leroyvn** [comment]
Closing (see 542d4acfa7625e64523a6b588badc45134650cab, which implements these changes and adds the corresponding documentation entry).
